File: opencode/session-picker.js
#!/usr/bin/env node

import { accessSync, constants, realpathSync } from "node:fs";
import { homedir } from "node:os";
import { basename, delimiter, join } from "node:path";
import process from "node:process";
import readline from "node:readline";
import { spawnSync } from "node:child_process";
import { fileURLToPath } from "node:url";
import { DatabaseSync } from "node:sqlite";

const RESET = "\u001b[0m";

function findExecutable(command) {
  for (const directory of (process.env.PATH ?? "").split(delimiter)) {
    const path = join(directory, command);
    try {
      accessSync(path, constants.X_OK);
      return path;
    } catch {}
  }
  throw new Error(`Missing required command: ${command}`);
}

function shellQuote(value) {
  return `'${value.replaceAll("'", `'\\''`)}'`;
}

function previewSession(sessionId) {
  if (!/^ses_[a-zA-Z0-9]+$/.test(sessionId)) {
    throw new Error("Invalid session ID");
  }

  const dataHome = process.env.XDG_DATA_HOME ?? join(homedir(), ".local", "share");
  const database = new DatabaseSync(join(dataHome, "opencode", "opencode.db"), {
    readOnly: true,
  });

  try {
    // Keep the preview conversational by excluding tool calls and reasoning parts.
    const messages = database
      .prepare(`
        SELECT
          json_extract(message.data, '$.role') AS role,
          json_extract(part.data, '$.text') AS text
        FROM part
        JOIN message ON message.id = part.message_id
        WHERE part.session_id = ?
          AND json_extract(part.data, '$.type') = 'text'
          AND json_extract(message.data, '$.role') IN ('user', 'assistant')
          AND trim(json_extract(part.data, '$.text')) != ''
        ORDER BY part.time_created DESC
        LIMIT 12
      `)
      .all(sessionId)
      .reverse();

    for (const { role, text } of messages) {
      const user = role === "user";
      const heading = user ? "\u001b[1;33mUSER" : "\u001b[1;36mASSISTANT";
      const body = user ? "\u001b[38;5;223m" : "\u001b[38;5;153m";
      process.stdout.write(`\n${heading}${RESET}\n${body}${text}${RESET}\n`);
    }
  } finally {
    database.close();
  }
}

function cleanField(value) {
  return String(value ?? "").replace(/[\t\r\n]/g, " ");
}

function formatUpdated(updated) {
  const date = new Date(updated);
  const parts = [
    date.getFullYear(),
    String(date.getMonth() + 1).padStart(2, "0"),
    String(date.getDate()).padStart(2, "0"),
    String(date.getHours()).padStart(2, "0"),
    String(date.getMinutes()).padStart(2, "0"),
  ];
  return `${parts[0]}-${parts[1]}-${parts[2]} ${parts[3]}:${parts[4]}`;
}

function listSessions(opencode) {
  const result = spawnSync(opencode, ["session", "list", "--format", "json", "-n", "100"], {
    encoding: "utf8",
  });
  if (result.error) throw result.error;
  if (result.status !== 0) {
    throw new Error(result.stderr.trim() || "Failed to list OpenCode sessions");
  }

  return JSON.parse(result.stdout).map((session) =>
    [
      cleanField(session.id),
      formatUpdated(session.updated),
      cleanField(session.title),
      cleanField(basename(session.directory)),
    ].join("\t"),
  );
}

function chooseSession(fzf, sessions) {
  const script = realpathSync(fileURLToPath(import.meta.url));
  const preview = `${shellQuote(process.execPath)} ${shellQuote(script)} --preview {1}`;
  const result = spawnSync(
    fzf,
    [
      "--ansi",
      "--delimiter=\t",
      "--with-nth=2..",
      "--layout=reverse",
      "--header=enter: select | esc: cancel",
      "--prompt=Sessions> ",
      `--preview=${preview}`,
      "--preview-window=right:60%:wrap",
    ],
    { input: `${sessions.join("\n")}\n`, encoding: "utf8", stdio: ["pipe", "pipe", "inherit"] },
  );
  if (result.error) throw result.error;
  return result.status === 0 ? result.stdout.trimEnd() : "";
}

async function chooseAction() {
  readline.emitKeypressEvents(process.stdin);
  if (!process.stdin.isTTY || typeof process.stdin.setRawMode !== "function") {
    throw new Error("Session action requires an interactive terminal");
  }

  process.stdout.write("\nenter: continue | f: fork | esc: cancel ");
  process.stdin.setRawMode(true);
  process.stdin.resume();

  try {
    while (true) {
      const key = await new Promise((resolve) => process.stdin.once("keypress", (_, value) => resolve(value)));
      if (key.name === "return") return "continue";
      if (key.name === "f") return "fork";
      if (key.name === "escape" || (key.ctrl && key.name === "c")) return "cancel";
    }
  } finally {
    process.stdin.setRawMode(false);
    process.stdin.pause();
    process.stdout.write("\n");
  }
}

async function main() {
  if (process.argv[2] === "--preview") {
    previewSession(process.argv[3] ?? "");
    return;
  }

  if (typeof process.execve !== "function") {
    throw new Error("Node.js 22.15 or newer is required");
  }

  const opencode = findExecutable("opencode");
  const fzf = findExecutable("fzf");
  const sessions = listSessions(opencode);
  if (sessions.length === 0) {
    throw new Error("No OpenCode sessions found for this project.");
  }

  const selection = chooseSession(fzf, sessions);
  if (!selection) return;

  const action = await chooseAction();
  if (action === "cancel") return;

  const sessionId = selection.split("\t", 1)[0];
  const args = ["opencode", "--session", sessionId];
  if (action === "fork") args.push("--fork");

  // Replace the wrapper process so the selected OpenCode TUI owns the terminal directly.
  process.execve(opencode, args, process.env);
}

main().catch((error) => {
  process.stderr.write(`${error.message}\n`);
  process.exitCode = 1;
});

File: opencode/session-picker.py
#!/usr/bin/env python3

import argparse
import os
import shlex
import shutil
import sqlite3
import subprocess
import sys
from datetime import datetime
from pathlib import Path
from urllib.parse import quote


def database_uri() -> str:
    data_home = Path(os.environ.get("XDG_DATA_HOME", Path.home() / ".local/share"))
    db_path = data_home / "opencode" / "opencode.db"
    return f"file:{quote(str(db_path), safe='/')}?mode=ro"


def preview_session(session_id: str) -> None:
    # Keep the preview conversational by excluding tool calls and reasoning parts.
    with sqlite3.connect(database_uri(), uri=True) as connection:
        rows = connection.execute(
            """
            SELECT
              json_extract(message.data, '$.role') AS role,
              json_extract(part.data, '$.text') AS text
            FROM part
            JOIN message ON message.id = part.message_id
            WHERE part.session_id = ?
              AND json_extract(part.data, '$.type') = 'text'
              AND json_extract(message.data, '$.role') IN ('user', 'assistant')
              AND trim(json_extract(part.data, '$.text')) != ''
            ORDER BY part.time_created DESC
            LIMIT 12
            """,
            (session_id,),
        ).fetchall()

    for role, text in reversed(rows):
        if role == "user":
            print(f"\n\033[1;33mUSER\033[0m\n\033[38;5;223m{text}\033[0m")
        else:
            print(f"\n\033[1;36mASSISTANT\033[0m\n\033[38;5;153m{text}\033[0m")


def list_sessions(query: str) -> list[str]:
    # A separate clone with the same origin shares sessions in OpenCode, but this
    # path-based lookup will not match until OpenCode registers it as a sandbox.
    try:
        result = subprocess.run(
            [
                "git",
                "rev-parse",
                "--path-format=absolute",
                "--git-common-dir",
                "--show-toplevel",
            ],
            stdout=subprocess.PIPE,
            stderr=subprocess.DEVNULL,
            text=True,
        )
    except FileNotFoundError:
        result = None

    if result is not None and result.returncode == 0:
        common_directory, checkout = result.stdout.splitlines()
        # A linked worktree's common Git directory lives under the primary worktree.
        worktree = Path(common_directory).parent
        scope = """
          project.worktree IN (?, ?)
          OR EXISTS (
            SELECT 1 FROM json_each(project.sandboxes) WHERE value IN (?, ?)
          )
        """
        parameters = (str(worktree), str(checkout), str(worktree), str(checkout))
    else:
        scope = "project.id = ?"
        parameters = ("global",)

    query_filter = ""
    limit = "LIMIT 100"
    if query:
        query_filter = """
          AND EXISTS (
            SELECT 1 FROM part
            WHERE part.session_id = session.id
              AND instr(lower(part.data), lower(?)) > 0
          )
        """
        parameters += (query,)
        limit = ""

    with sqlite3.connect(database_uri(), uri=True) as connection:
        sessions = connection.execute(
            f"""
            SELECT session.id, session.time_updated, session.title, session.directory
            FROM session
            JOIN project ON project.id = session.project_id
            WHERE session.parent_id IS NULL
              AND session.time_archived IS NULL
              AND ({scope})
              {query_filter}
            ORDER BY session.time_updated DESC
            {limit}
            """,
            parameters,
        ).fetchall()

    rows = []

    for session_id, time_updated, session_title, session_directory in sessions:
        updated = datetime.fromtimestamp(time_updated / 1000)
        title = session_title.translate(
            str.maketrans({"\t": " ", "\r": " ", "\n": " "})
        )
        directory = Path(session_directory).name
        # The hidden first TSV field carries the session ID through selection.
        rows.append(f"{session_id}\t{updated:%Y-%m-%d %H:%M}\t{title}\t{directory}")

    return rows


def choose_session(rows: list[str]) -> tuple[str, str] | None:
    # fzf owns the list, filtering, and live right-hand preview.
    script_path = Path(__file__).resolve()
    if rows:
        header = "enter: continue | alt-enter: fork | ctrl-n: new | esc: cancel"
        selection_bindings = [
            "--bind=enter:print(continue)+accept",
            "--bind=alt-enter:print(fork)+accept",
        ]
    else:
        header = "ctrl-n: new | esc: cancel"
        selection_bindings = ["--bind=enter:ignore"]

    result = subprocess.run(
        [
            "fzf",
            "--ansi",
            "--delimiter=\t",
            "--with-nth=2..",
            "--layout=reverse",
            f"--header={header}",
            "--prompt=Sessions> ",
            *selection_bindings,
            "--bind=ctrl-n:print(new)+accept",
            f"--preview={shlex.quote(str(script_path))} --internal-preview {{1}}",
            "--preview-window=right:60%:wrap",
        ],
        input="\n".join(rows) if rows else "\t-- No sessions found --",
        stdout=subprocess.PIPE,
        text=True,
    )
    if result.returncode != 0 or not result.stdout.strip():
        return None
    action, _, selection = result.stdout.partition("\n")
    return action, selection.partition("\t")[0]


def main() -> int:
    parser = argparse.ArgumentParser()
    mode = parser.add_mutually_exclusive_group()
    mode.add_argument("-q", "--query", default="", metavar="TEXT")
    mode.add_argument("--internal-preview", metavar="SESSION_ID", help=argparse.SUPPRESS)
    arguments = parser.parse_args()

    # fzf reinvokes this script in an internal mode as the highlighted session changes.
    if arguments.internal_preview is not None:
        preview_session(arguments.internal_preview)
        return 0

    for command in ("opencode", "fzf"):
        if shutil.which(command) is None:
            print(f"Missing required command: {command}", file=sys.stderr)
            return 1

    sessions = list_sessions(arguments.query)
    selection = choose_session(sessions)
    if selection is None:
        return 0
    action, session_id = selection

    arguments = ["opencode"]
    if action != "new":
        arguments.extend(("--session", session_id))
        if action == "fork":
            arguments.append("--fork")
    # Replace the wrapper process so the selected OpenCode TUI owns the terminal directly.
    os.execvp(arguments[0], arguments)


if __name__ == "__main__":
    try:
        sys.exit(main())
    except (OSError, sqlite3.Error) as error:
        print(f"opencode-session-picker: {error}", file=sys.stderr)
        sys.exit(1)
